### PR TITLE
Expose `algorithm` field on `PublicKey`

### PR DESCRIPTION
**Pull request by rickvanprim**
Fixes https://github.com/rustls/rcgen/issues/280.

### CODE REVIEW

File: rcgen/src/csr.rs [hunk 1]
```diff
@@ -16,7 +16,8 @@ use crate::{DistinguishedName, SanType};
 #[derive(Debug, PartialEq, Eq, Hash)]
 pub struct PublicKey {
 	raw: Vec<u8>,
-	alg: &'static SignatureAlgorithm,
+	/// The algorithm used to generate the public key.
```
Comment (cpu) [inline rcgen/src/csr.rs:19]:
```suggestion
	/// The algorithm used to generate the public key and sign the CSR.
```

**Review by cpu — APPROVED** [review]
This seems reasonable to me. AIUI this is a capability callers had before  https://github.com/rustls/rcgen/pull/205

File: rcgen/src/csr.rs (hunk 1 above)
Comment (rickvanprim) [reply rcgen/src/csr.rs]: Done

Comment (cpu) [reply rcgen/src/csr.rs]: Thanks!

**Comment by rickvanprim** [comment]
> This seems reasonable to me. AIUI this is a capability callers had before #205

My code on the older version of `rcgen` did do the algorithm check as it was exposed as part of `CertificateParams` 🙂

**Review by djc — APPROVED** [review]

**Review by est31 — CHANGES_REQUESTED** [review]
I think it would be better to add it as a function on `PublicKey` instead, then `mut` access doesn't allow you to generate invalid configurations.

Also, I wonder whether it makes sense to additionally expose the `PublicKeyData` trait to the public (I don't like APIs that require you to import a trait, ideally the functions are available both through traits and inherent functions, unless it is a lot of API surface that's covered). Maybe that's a separate discussion though.

**Comment by cpu** [comment]
> I think it would be better to add it as a function on PublicKey instead, then mut access doesn't allow you to generate invalid configurations.

Sounds reasonable to me.

> Also, I wonder whether it makes sense to additionally expose the PublicKeyData trait to the public

I like this direction less personally. I think this trait is a bit awkward (but have a hard time articulating why). I think exposing it in the existing API will make it harder to smooth out any rough edges.

**Comment by rickvanprim** [comment]
@est31 / @cpu, I changed to a method on `PublicKey`, which is a copy paste of the method for `PublicKeyData`.  I could make the `PublicKeyData` method call the method on `PublicKey` as well if that's preferred, but given how they're both trivial accessors, I'm not sure that serves much point.  Please let me know if any further changes are desired 🙂

One small discrepancy I just noticed while adding this to `test_csr_roundtrip()` is `KeyPair::algorithm()` vs `PublicKeyData::alg()`.

**Review by est31 — APPROVED** [review]
Thanks!

File: rcgen/src/csr.rs [hunk 2]
```diff
@@ -19,6 +19,13 @@ pub struct PublicKey {
 	alg: &'static SignatureAlgorithm,
 }
 
+impl PublicKey {
+	/// The algorithm used to generate the public key and sign the CSR.
+	pub fn alg(&self) -> &SignatureAlgorithm {
```
Comment (cpu) [inline rcgen/src/csr.rs:24]: For a new fn where we have some flexibility I think `algorithm` would be a better name and will resolve the mismatch you called out in your unit test w.r.t the `assert_eq!`. Let's do that?

Comment (rickvanprim) [reply rcgen/src/csr.rs]: Done 🙂

**Review by cpu — APPROVED** [review]
Thanks!

**Pull request merged**